cilium-dbg/cmd/shell.go

```diff
@@ -0,0 +1,179 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package cmd
+
+import (
+	"bufio"
+	"fmt"
+	"io"
+	"net"
+	"os"
+	"os/signal"
+	"strings"
+	"time"
+
+	"github.com/cilium/hive/script"
+	"github.com/spf13/cobra"
+	"golang.org/x/term"
+
+	"github.com/cilium/cilium/pkg/defaults"
+	"github.com/cilium/cilium/pkg/version"
+)
+
+var shellCmd = &cobra.Command{
+	Use:   "shell [command] [args]...",
+	Short: "Connect to the Cilium shell",
+	Run:   shell,
+}
+
+var stdReadWriter = struct {
+	io.Reader
+	io.Writer
+}{
+	Reader: os.Stdin,
+	Writer: os.Stdout,
+}
+
+func init() {
+	RootCmd.AddCommand(shellCmd)
+}
+
+func dialShell() (net.Conn, error) {
+	var conn net.Conn
+	until := time.Now().Add(time.Minute)
+	for {
+		var err error
+		conn, err = net.Dial("unix", defaults.ShellSockPath)
+		if err == nil {
+			break
+		}
+		if time.Now().After(until) {
+			return nil, fmt.Errorf("Timed out dialing %q\n", defaults.ShellSockPath)
+		}
+		fmt.Fprintf(os.Stderr, "Dial(%q): %s. Retrying...\n", defaults.ShellSockPath, err)
+		time.Sleep(time.Second)
+	}
+	return conn, nil
+}
+
+func shellExchange(w io.Writer, format string, args ...any) error {
+	conn, err := dialShell()
+	if err != nil {
+		return err
+	}
+	defer conn.Close()
+	_, err = fmt.Fprintf(conn, format+"\nexit\n", args...)
+	if err != nil {
+		return err
+	}
+	bio := bufio.NewReader(conn)
+	for {
+		line, _, err := bio.ReadLine()
+		if err != nil {
+			return nil
+		}
+		switch string(line) {
+		case "<<end>>":
+			return nil
+		case "[stdout]":
+		default:
+			w.Write(line)
+			w.Write(newline)
+		}
+	}
+
+}
+
+func shell(cmd *cobra.Command, args []string) {
+	if len(args) > 0 {
+		shellExchange(os.Stdout, "%s", strings.Join(args, " "))
+	} else {
+		interactiveShell()
+	}
+}
+
+func interactiveShell() {
+	// Try to set the terminal to raw mode (so that cursor keys work etc.)
+	restore, err := script.MakeRaw(0)
+	if err != nil {
+		panic(err)
```

Comment:
better return the error from here and handle in shell()?

Reply:
Since it's non-fatal I'll just print the error to stderr and continue. Effect is that cursor keys etc. won't work.